File: Beginner/Python/Tree/IsValidBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
            :type root: TreeNode
            :rtype: bool
            """
        return self.helper(root, float("-inf"), float("inf"))
    
    def helper(self, root, left, right):
        if root == None:
            return True
        elif not left < root.val < right:
            return False
        else:
            return self.helper(root.left, left, root.val) and self.helper(root.right, root.val, right)

File: Beginner/Java/Tree/IsValidBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null || (root.left == null && root.right == null))
            return true;
        if (isValidBST(root.left) && isValidBST(root.right)) {
            TreeNode pl = root.left;
            TreeNode pr = root.right;
            while (pl != null && pl.right != null)
                pl = pl.right;
            while (pr != null && pr.left != null)
                pr = pr.left;
            if ((pl == null || pl.val < root.val) && (pr == null || pr.val > root.val))
                return true;
        }
        return false;
    }
}
